---
title: "R Notebook"
output:
  html_document:
    df_print: paged
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

```{r}
library("DESeq2")
library("tximport")
library("GenomicFeatures")
library("biomaRt")
library("reshape")
library(reshape2)
library(ggplot2)
library(pheatmap)
library(scater)
library("RColorBrewer")
library(dplyr)
library(tidyr)
library(tibble)
library("org.Mm.eg.db")
library(RUVSeq)
library("IHW")
library(Biobase)
library(preprocessCore)
library(EDASeq)
library(fgsea)

```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

Load sample metadata
```{r}
sample_metadata_all <- read.csv("sample_metadata_180619.csv", header = T)
```

Build transcript <> gene table (+ ERCC)
```{r}
txdb <- makeTxDbFromGFF('/zfs/referenceGenomes/mouse/Mus_musculus.GRCm38.96.gtf')
k <- keys(txdb, keytype = "TXNAME")
tx2gene <- AnnotationDbi::select(txdb, k, "GENEID", "TXNAME")
ERCC92 <- read.csv("ERCC92_names.txt", header = F)
ERCC92$V2 <- ERCC92$V1
colnames(ERCC92) <- c("TXNAME", "GENEID")
tx2gene_ERCC <- bind_rows(tx2gene, ERCC92)
```

Parameters: Comparison 1 -> Axon IP vs Midbrain IP
```{r}
path <- "/zfs/analysis/active_r_projects/trap_kall_protcod_ercc/"

sample_metadata <- sample_metadata_all %>% data.frame() %>%
                    as_tibble() %>% filter(ip == 'ip' & batch_trap %in% c('b3'))

files <- file.path(path, sample_metadata$sample_code, "abundance.h5")
names(files) <- sample_metadata$sample_name

design <- ~ collection_day + age + axon

min_counts = 100

contrast <- c("axon", "mb", "axon")

alpha = 0.01

res_lfc = 0 #lfc threshold for DESeq2 results functions

lfc = 0 #lfc threshold for post-results filtering genes

replicate_replace = Inf
```

Import kallisto output
```{r}
txi <- tximport(files, type = "kallisto", tx2gene = tx2gene_ERCC, ignoreTxVersion = T)
```
Create DESeq2 object
```{r}
dds <- DESeqDataSetFromTximport(txi, colData = sample_metadata, design = design)
```

Generate annotations
```{r}
# mart <- useMart(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
# anno <- getBM(attributes = c("ensembl_gene_id", "external_gene_name", "description"), filters = "ensembl_gene_id", values = rownames(counts(dds)), mart = mart, uniqueRows = T)
# save(anno, file = 'anno.RData')

load('anno.RData')
```

Filter to remove low counts
```{r}
keep_feature <- rowMeans(counts(dds)) >= min_counts
dds <- dds[keep_feature, ]
dim(dds)
summary(counts(dds))
dds <- DESeq(dds, minReplicatesForReplace = replicate_replace)
contrast_name = paste('dds', contrast[2], 'vs', contrast[3], sep = '')
dds_axonip_vs_midbrainip <- dds
```

List contrasts to set lfcShrink coefficient
```{r}
resultsNames(dds)
```

DESeq2 result
```{r}
res <- results(dds, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
summary(res)
```

'Apeglm' shrinkage result
```{r}
shrinkage_coef = 7 #run 'resultsNames(dds)' to list coefficients

res_shrunken <- lfcShrink(dds, coef = shrinkage_coef, type = "apeglm", lfcThreshold = res_lfc, res = res)
summary(res_shrunken)
```

MA-plot
```{r}
DESeq2::plotMA(res_shrunken, ylim=c(-5,5))
```

Independent hypothesis weighting (using baseMean to adjust p-values)
```{r}
deRes <- as.data.frame(res)
ihwRes <- ihw(pvalue ~ baseMean,  data = deRes, alpha = alpha)
resIHW <- results(dds, filterFun=ihw, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
summary(resIHW)
```
Extract significant results with annotation information. Choose results output:
```{r}
res = resIHW

res_tb <- res %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()

sig <- res_tb %>%
  filter(padj < alpha & abs(log2FoldChange) > lfc)

sig <- merge(as.data.frame(sig), anno, by.x = c("gene"), by.y = c("ensembl_gene_id"))
```

Create Axon IP vs Midbrain IP significant object
```{r}
axonip_vs_midbrainip <- sig
```

Parameters: Comparison 2 -> Axon UB vs Midbrain UB
```{r}
sample_metadata <- sample_metadata_all %>% data.frame() %>%
                    as_tibble() %>% filter(ip == 'input' & batch_trap %in% c('b2', 'b3'))

files <- file.path(path, sample_metadata$sample_code, "abundance.h5")
names(files) <- sample_metadata$sample_name

design <- ~ batch_trap + age + axon

contrast <- c("axon", "mb", "axon")

min_counts = 100
```

Import kallisto output
```{r}
txi <- tximport(files, type = "kallisto", tx2gene = tx2gene_ERCC, ignoreTxVersion = T)
```

Create DESeq2 object and filter to remove low counts
```{r}
dds <- DESeqDataSetFromTximport(txi, colData = sample_metadata, design = design)
keep_feature <- rowMeans(counts(dds)) >= min_counts
dds <- dds[keep_feature, ]
dim(dds)
summary(counts(dds))
dds <- DESeq(dds, minReplicatesForReplace = replicate_replace)
dds_axonub_vs_midbrainub <- dds
```

List contrasts to set lfcShrink coefficient
```{r}
resultsNames(dds)
```

DESeq2 result
```{r}
res <- results(dds, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
summary(res)
```

'Apeglm' shrinkage result
```{r}
shrinkage_coef = 4 #run 'resultsNames(dds)' to list coefficients

res_shrunken <- lfcShrink(dds, coef = shrinkage_coef, type = "apeglm", lfcThreshold = res_lfc, res = res)
summary(res_shrunken)
```

MA-plot
```{r}
DESeq2::plotMA(res_shrunken, ylim=c(-5,5))
```

Independent hypothesis weighting (using baseMean to adjust p-values)
```{r}
deRes <- as.data.frame(res)
ihwRes <- ihw(pvalue ~ baseMean,  data = deRes, alpha = alpha)
resIHW <- results(dds, filterFun=ihw, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
summary(resIHW)
```

Extract significant results with annotation information. Choose results output:
```{r}
res = resIHW

res_tb <- res %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()

sig <- res_tb %>%
  filter(padj < alpha & abs(log2FoldChange) > lfc)

sig <- merge(as.data.frame(sig), anno, by.x = c("gene"), by.y = c("ensembl_gene_id"))
```

Create Axon UB vs Midbrain UB significant object
```{r}
axonub_vs_midbrainub <- sig
```

Now a comparison of IP and UB to identify genes enriched in either region. These genes

Midbrain IP vs Input with 0 log-fold threshold
```{r}
sample_metadata <- sample_metadata_all %>% data.frame() %>%
                    as_tibble() %>% filter(region == 'mb' & batch_trap %in% c('b2', 'b3'))

files <- file.path(path, sample_metadata$sample_code, "abundance.h5")
names(files) <- sample_metadata$sample_name

design <- ~ batch_trap + age + ip

contrast <- c("ip", "ip", "input")

lfc <- 0

min_counts <- 10

```

```{r}
txi <- tximport(files, type = "kallisto", tx2gene = tx2gene_ERCC, ignoreTxVersion = T)
```

Create DESeq2 object and filter to remove low counts
```{r}
dds <- DESeqDataSetFromTximport(txi, colData = sample_metadata, design = design)
keep_feature <- rowMeans(counts(dds)) >= min_counts
dds <- dds[keep_feature, ]
dim(dds)
summary(counts(dds))
dds <- DESeq(dds, minReplicatesForReplace = replicate_replace)
dds_midbrainip_vs_midbrainub <- dds
```

List contrasts to set lfcShrink coefficient
```{r}
resultsNames(dds)
```

DESeq2 result
```{r}
res <- results(dds, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
summary(res)
```
'Apeglm' shrinkage result
```{r}
shrinkage_coef = 4 #run 'resultsNames(dds)' to list coefficients

res_shrunken <- lfcShrink(dds, coef = shrinkage_coef, type = "apeglm", lfcThreshold = res_lfc, res = res)
summary(res_shrunken)
```

MA-plot
```{r}
DESeq2::plotMA(res_shrunken, ylim=c(-5,5))
```

Independent hypothesis weighting (using baseMean to adjust p-values)
```{r}
deRes <- as.data.frame(res)
ihwRes <- ihw(pvalue ~ baseMean,  data = deRes, alpha = alpha)
resIHW <- results(dds, filterFun=ihw, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
summary(resIHW)
```

Extract significant results with annotation information. Choose results output:
```{r}
res = resIHW

res_tb <- res %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()

sig <- res_tb %>%
  filter(padj < alpha & abs(log2FoldChange) > lfc)

sig <- merge(as.data.frame(sig), anno, by.x = c("gene"), by.y = c("ensembl_gene_id"))
```

Create Axon IP vs Midbrain IP significant object
```{r}
midbrainip_vs_input <- sig
```

Axon IP vs Unbound
```{r}
sample_metadata <- sample_metadata_all %>% data.frame() %>%
                    as_tibble() %>% filter(axon == 'axon' & batch_trap %in% c('b3'))

files <- file.path(path, sample_metadata$sample_code, "abundance.h5")
names(files) <- sample_metadata$sample_name

design <- ~ age + ip

contrast <- c("ip", "ip", "input")

lfc <- 0

min_counts <- 10

```

```{r}
txi <- tximport(files, type = "kallisto", tx2gene = tx2gene_ERCC, ignoreTxVersion = T)
```

Create DESeq2 object and filter to remove low counts
```{r}
dds <- DESeqDataSetFromTximport(txi, colData = sample_metadata, design = design)
keep_feature <- rowMeans(counts(dds)) >= min_counts
dds <- dds[keep_feature, ]
dim(dds)
summary(counts(dds))
dds <- DESeq(dds, minReplicatesForReplace = replicate_replace)
dds_axonip_vs_axonub <- dds
```

List contrasts to set lfcShrink coefficient
```{r}
resultsNames(dds)
```

DESeq2 result
```{r}
res <- results(dds, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
summary(res)
```
'Apeglm' shrinkage result
```{r}
shrinkage_coef = 3 #run 'resultsNames(dds)' to list coefficients

res_shrunken <- lfcShrink(dds, coef = shrinkage_coef, type = "apeglm", lfcThreshold = res_lfc, res = res)
summary(res_shrunken)
```

MA-plot
```{r}
DESeq2::plotMA(res_shrunken, ylim=c(-5,5))
```

Independent hypothesis weighting (using baseMean to adjust p-values)
```{r}
deRes <- as.data.frame(res)
ihwRes <- ihw(pvalue ~ baseMean,  data = deRes, alpha = alpha)
resIHW <- results(dds, filterFun=ihw, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
summary(resIHW)
```

Extract significant results with annotation information. Choose results output:
```{r}
res = resIHW

res_tb <- res %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()

sig <- res_tb %>%
  filter(padj < alpha & abs(log2FoldChange) > lfc)

sig <- merge(as.data.frame(sig), anno, by.x = c("gene"), by.y = c("ensembl_gene_id"))
```

Create Axon IP vs Input significant object
```{r}
axonip_vs_input <- sig
```

Identify Axon and Midbrain enriched and specific genes
```{r}
axon_translated <- axonip_vs_input %>% filter(log2FoldChange > 0)
midbrain_translated <- midbrainip_vs_input %>% filter(log2FoldChange > 0)

axon_only <- axon_translated %>% filter(!(gene %in% midbrain_translated$gene))
midbrain_only <- midbrain_translated %>% filter(!(gene %in% axon_translated$gene))

# axon_up_ub <- axonub_vs_midbrainub %>% filter(log2FoldChange < 0)
# axon_up_ip0 <- axonip_vs_midbrainip %>% filter(log2FoldChange < 0)
# axon_up_untranslated_ub <- axon_up_ub %>% filter(!(gene %in% axon_translated$gene))
# axon_up_untranslated_ip <- axon_up_ip0 %>% filter(!(gene %in% axon_translated$gene))
# axon_up_ip1 <- axon_up_ip0 %>% filter(!(gene %in% axon_up_untranslated_ub$gene))
# axon_up_ip <- axon_up_ip1 %>% filter(!(gene %in% axon_up_untranslated_ip$gene))
# axon_only <- axon_up_ip %>% filter(!(gene %in% midbrain_translated$gene))
# 
# midbrain_up_ub <- axonub_vs_midbrainub %>% filter(log2FoldChange > 0)
# midbrain_up_ip0 <- axonip_vs_midbrainip %>% filter(log2FoldChange > 0)
# midbrain_up_untranslated_ub <- midbrain_up_ub %>% filter(!(gene %in% midbrain_translated$gene))
# midbrain_up_untranslated_ip <- midbrain_up_ip0 %>% filter(!(gene %in% midbrain_translated$gene))
# midbrain_up_ip1 <- midbrain_up_ip0 %>% filter(!(gene %in% midbrain_up_untranslated_ub$gene))
# midbrain_up_ip <- midbrain_up_ip1 %>% filter(!(gene %in% midbrain_up_untranslated_ip$gene))
# midbrain_only <- midbrain_up_ip %>% filter(!(gene %in% axon_translated$gene))

axon_and_midbrain_translated <- axon_translated %>% filter(gene %in% midbrain_translated$gene)
```

Example of case where neighbouring expression of a midbrain-only transcript is counteracted by excluding UB Axon vs UB midbrain DEGs that are not in the IP/INPUT (translated) lists.

```{r}
gene = 'ENSMUSG00000029491'

plotCounts(dds_axonip_vs_axonub, gene=gene, intgroup="ip")
plotCounts(dds_midbrainip_vs_midbrainub, gene=gene, intgroup="ip")
plotCounts(dds_axonip_vs_midbrainip, gene=gene, intgroup="axon")
plotCounts(dds_axonub_vs_midbrainub, gene=gene, intgroup="axon")
plotCounts(dds_vsip_vs_vsub, gene=gene, intgroup="ip")
plotCounts(dds_dsip_vs_dsub, gene=gene, intgroup="ip")
```

Dorsal vs Ventral IP
```{r}
sample_metadata <- sample_metadata_all %>% data.frame() %>%
                    as_tibble() %>% filter(axon == 'axon' & ip == 'ip')

files <- file.path(path, sample_metadata$sample_code, "abundance.h5")
names(files) <- sample_metadata$sample_name

design <- ~ collection_day + age + region

contrast <- c("region", "vs", "ds")

lfc <- 0

min_counts <- 100

```

```{r}
txi <- tximport(files, type = "kallisto", tx2gene = tx2gene_ERCC, ignoreTxVersion = T)
```

Create DESeq2 object and filter to remove low counts
```{r}
dds <- DESeqDataSetFromTximport(txi, colData = sample_metadata, design = design)
keep_feature <- rowMeans(counts(dds)) >= min_counts
dds <- dds[keep_feature, ]
dim(dds)
summary(counts(dds))
dds <- DESeq(dds, minReplicatesForReplace = replicate_replace)
dds_dsip_vs_vsip <- dds
```

List contrasts to set lfcShrink coefficient
```{r}
resultsNames(dds)
```

DESeq2 result
```{r}
res <- results(dds, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
summary(res)
```
'Apeglm' shrinkage result
```{r}
shrinkage_coef = 7 #run 'resultsNames(dds)' to list coefficients

res_shrunken <- lfcShrink(dds, coef = shrinkage_coef, type = "apeglm", lfcThreshold = res_lfc, res = res)
summary(res_shrunken)
```

MA-plot
```{r}
DESeq2::plotMA(res_shrunken, ylim=c(-5,5))
```

Independent hypothesis weighting (using baseMean to adjust p-values)
```{r}
deRes <- as.data.frame(res)
ihwRes <- ihw(pvalue ~ baseMean,  data = deRes, alpha = alpha)
resIHW <- results(dds, filterFun=ihw, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
summary(resIHW)
```

Extract significant results with annotation information. Choose results output:
```{r}
res = resIHW

res_tb <- res %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()

sig <- res_tb %>%
  filter(padj < alpha & abs(log2FoldChange) > lfc)

sig <- merge(as.data.frame(sig), anno, by.x = c("gene"), by.y = c("ensembl_gene_id"))
```

Create vs IP vs ds IP significant object
```{r}
vsip_vs_dsip <- sig
```

Dorsal UB vs Ventral UB
```{r}
sample_metadata <- sample_metadata_all %>% data.frame() %>%
                    as_tibble() %>% filter(axon == 'axon' & ip == 'input')

files <- file.path(path, sample_metadata$sample_code, "abundance.h5")
names(files) <- sample_metadata$sample_name

design <- ~ age + region

contrast <- c("region", "vs", "ds")

lfc <- 0

min_counts <- 100

```

```{r}
txi <- tximport(files, type = "kallisto", tx2gene = tx2gene_ERCC, ignoreTxVersion = T)
```

Create DESeq2 object and filter to remove low counts
```{r}
dds <- DESeqDataSetFromTximport(txi, colData = sample_metadata, design = design)
keep_feature <- rowMeans(counts(dds)) >= min_counts
dds <- dds[keep_feature, ]
dim(dds)
summary(counts(dds))
dds <- DESeq(dds, minReplicatesForReplace = replicate_replace)

dds_dsub_vs_vsub <- dds
```

List contrasts to set lfcShrink coefficient
```{r}
resultsNames(dds)
```

DESeq2 result
```{r}
res <- results(dds, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
summary(res)
```
'Apeglm' shrinkage result
```{r}
shrinkage_coef = 3 #run 'resultsNames(dds)' to list coefficients

res_shrunken <- lfcShrink(dds, coef = shrinkage_coef, type = "apeglm", lfcThreshold = res_lfc, res = res)
summary(res_shrunken)
```

MA-plot
```{r}
DESeq2::plotMA(res_shrunken, ylim=c(-5,5))
```

Independent hypothesis weighting (using baseMean to adjust p-values)
```{r}
deRes <- as.data.frame(res)
ihwRes <- ihw(pvalue ~ baseMean,  data = deRes, alpha = alpha)
resIHW <- results(dds, filterFun=ihw, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
summary(resIHW)
```

Extract significant results with annotation information. Choose results output:
```{r}
res = resIHW

res_tb <- res %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()

sig <- res_tb %>%
  filter(padj < alpha & abs(log2FoldChange) > lfc)

sig <- merge(as.data.frame(sig), anno, by.x = c("gene"), by.y = c("ensembl_gene_id"))
```

Create Dorsal UB vs Ventral UB significant object
```{r}
vsub_vs_dsub <- sig
```

Identify DS/VS enriched genes. Note 'ds/vs_up_axon_only' does not mean they are only expressed in ds or vs. It means they are expressed axon only, but a comparison of ds/vs IP vs input has not been done to confirm whether truly region specific yet.
```{r}
vs_up_ip0 <- vsip_vs_dsip %>% filter(log2FoldChange > 0)
vs_up_ub <- vsub_vs_dsub %>% filter(log2FoldChange > 0)
vs_up_untranslated_ub <- vs_up_ub %>% filter(!(gene %in% axon_translated$gene))
vs_up_untranslated_ip <- vs_up_ip0 %>% filter(!(gene %in% axon_translated$gene))
vs_up_ip1 <- vs_up_ip0 %>% filter(!(gene %in% vs_up_untranslated_ub$gene))
vs_up_ip <- vs_up_ip1 %>% filter(!(gene %in% vs_up_untranslated_ip$gene))

vs_up_axon_only <- vs_up_ip %>% filter(gene %in% axon_only$gene)
  
ds_up_ip0 <- vsip_vs_dsip %>% filter(log2FoldChange < 0)
ds_up_ub <- vsub_vs_dsub %>% filter(log2FoldChange < 0)
ds_up_untranslated_ub <- ds_up_ub %>% filter(!(gene %in% axon_translated$gene))
ds_up_untranslated_ip <- ds_up_ip0 %>% filter(!(gene %in% axon_translated$gene))
ds_up_ip1 <- ds_up_ip0 %>% filter(!(gene %in% ds_up_untranslated_ub$gene))
ds_up_ip <- ds_up_ip1 %>% filter(!(gene %in% ds_up_untranslated_ip$gene))

ds_up_axon_only <- ds_up_ip %>% filter(gene %in% axon_only$gene)
```

DS IP vs DS Input
```{r}
sample_metadata <- sample_metadata_all %>% data.frame() %>%
  as_tibble() %>% filter(region == 'ds')

files <- file.path(path, sample_metadata$sample_code, "abundance.h5")
names(files) <- sample_metadata$sample_name

design <- ~ age + ip

contrast <- c("ip", "ip", "input")

lfc <- 0

min_counts <- 10
```

```{r}
txi <- tximport(files, type = "kallisto", tx2gene = tx2gene_ERCC, ignoreTxVersion = T)
```

Create DESeq2 object and filter to remove low counts
```{r}
dds <- DESeqDataSetFromTximport(txi, colData = sample_metadata, design = design)
keep_feature <- rowMeans(counts(dds)) >= min_counts
dds <- dds[keep_feature, ]
dim(dds)
summary(counts(dds))
dds <- DESeq(dds, minReplicatesForReplace = replicate_replace)

dds_dsip_vs_dsub <- dds
```

List contrasts to set lfcShrink coefficient
```{r}
resultsNames(dds)
```

DESeq2 result
```{r}
res <- results(dds, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
summary(res)
```
'Apeglm' shrinkage result
```{r}
shrinkage_coef = 3 #run 'resultsNames(dds)' to list coefficients

res_shrunken <- lfcShrink(dds, coef = shrinkage_coef, type = "apeglm", lfcThreshold = res_lfc, res = res)
summary(res_shrunken)
```

MA-plot
```{r}
DESeq2::plotMA(res_shrunken, ylim=c(-5,5))
```

Independent hypothesis weighting (using baseMean to adjust p-values)
```{r}
deRes <- as.data.frame(res)
ihwRes <- ihw(pvalue ~ baseMean,  data = deRes, alpha = alpha)
resIHW <- results(dds, filterFun=ihw, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
summary(resIHW)
```

Extract significant results with annotation information. Choose results output:
```{r}
res = resIHW

res_tb <- res %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()

sig <- res_tb %>%
  filter(padj < alpha & abs(log2FoldChange) > lfc)

sig <- merge(as.data.frame(sig), anno, by.x = c("gene"), by.y = c("ensembl_gene_id"))
```

```{r}
dsip_vs_input <- sig
```


VS IP vs VS Input
```{r}
sample_metadata <- sample_metadata_all %>% data.frame() %>%
  as_tibble() %>% filter(region == 'vs')

files <- file.path(path, sample_metadata$sample_code, "abundance.h5")
names(files) <- sample_metadata$sample_name

design <- ~ age + ip

contrast <- c("ip", "ip", "input")

lfc <- 0

min_counts <- 10
```

```{r}
txi <- tximport(files, type = "kallisto", tx2gene = tx2gene_ERCC, ignoreTxVersion = T)
```

Create DESeq2 object and filter to remove low counts
```{r}
dds <- DESeqDataSetFromTximport(txi, colData = sample_metadata, design = design)
keep_feature <- rowMeans(counts(dds)) >= min_counts
dds <- dds[keep_feature, ]
dim(dds)
summary(counts(dds))
dds <- DESeq(dds, minReplicatesForReplace = replicate_replace)

dds_vsip_vs_vsub <- dds
```

List contrasts to set lfcShrink coefficient
```{r}
resultsNames(dds)
```

DESeq2 result
```{r}
res <- results(dds, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
summary(res)
```
'Apeglm' shrinkage result
```{r}
shrinkage_coef = 3 #run 'resultsNames(dds)' to list coefficients

res_shrunken <- lfcShrink(dds, coef = shrinkage_coef, type = "apeglm", lfcThreshold = res_lfc, res = res)
summary(res_shrunken)
```

MA-plot
```{r}
DESeq2::plotMA(res_shrunken, ylim=c(-5,5))
```

Independent hypothesis weighting (using baseMean to adjust p-values)
```{r}
deRes <- as.data.frame(res)
ihwRes <- ihw(pvalue ~ baseMean,  data = deRes, alpha = alpha)
resIHW <- results(dds, filterFun=ihw, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
summary(resIHW)
```

Extract significant results with annotation information. Choose results output:
```{r}
res = resIHW

res_tb <- res %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()

sig <- res_tb %>%
  filter(padj < alpha & abs(log2FoldChange) > lfc)

sig <- merge(as.data.frame(sig), anno, by.x = c("gene"), by.y = c("ensembl_gene_id"))
```

```{r}
vsip_vs_input <- sig
```

```{r}
ds_translated <- dsip_vs_input %>% filter(log2FoldChange > 0)
vs_translated <- vsip_vs_input %>% filter(log2FoldChange > 0)

ds_ip_only <- ds_up_axon_only %>% filter(!(gene %in% vsip_vs_input$gene))
ds_ip_only0 <- ds_translated %>% filter(!(gene %in% vs_translated$gene))
ds_ip_only <- ds_ip_only0 %>% filter(!(gene %in% midbrain_translated$gene))

vs_ip_only <- vs_up_axon_only %>% filter(!(gene %in% dsip_vs_input$gene))
vs_ip_only0 <- vs_translated %>% filter(!(gene %in% ds_translated$gene))
vs_ip_only <- vs_ip_only0 %>% filter(!(gene %in% midbrain_translated$gene))

```

```{r}
paste('There are ', nrow(midbrain_translated), ' translated genes in the midbrain', sep = '')
paste('There are ', nrow(axon_translated), ' translated genes in the axon', sep = '')
paste('There are ', nrow(axon_and_midbrain_translated), ' translated genes shared between axon and midbrain', sep = '')
paste('There are ', nrow(midbrain_only), ' translated genes only in the midbrain', sep = '')
paste('There are ', nrow(axon_only), ' translated genes only in the axon', sep = '')
paste('There are ', nrow(ds_ip_only), ' translated genes only in the dorsal striatal axon', sep = '')
paste('There are ', nrow(vs_ip_only), ' translated genes only in the ventral striatal axon', sep = '')
paste('No dorsal/striatal-specific genes are translated in the midbrain')
```

Export results
```{r}
write.csv(axon_translated, file = paste('axon_translated_', Sys.Date(), '.csv', sep = ''))
write.csv(midbrain_translated, file = paste('midbrain_translated_', Sys.Date(), '.csv', sep = ''))
write.csv(axon_only, file = paste('axon_only_', Sys.Date(), '.csv', sep = ''))
write.csv(midbrain_only, file = paste('midbrain_only_', Sys.Date(), '.csv', sep = ''))
write.csv(ds_ip_only, file = paste('ds_only_', Sys.Date(), '.csv', sep = ''))
write.csv(vs_ip_only, file = paste('vs_only_', Sys.Date(), '.csv', sep = ''))

axon_or_midbrain_translated <- bind_rows(axon_translated, midbrain_translated) %>% distinct(gene, .keep_all = TRUE)
write.csv(axon_or_midbrain_translated, file = paste('axon_or_midbrain_translated_', Sys.Date(), '.csv', sep = ''))
```

```{r}
calcium <- as_tibble(read.csv(file = 'gene_lists/axon_ontologies/calcium.csv', header = T))
calcium <- unique(c(as.character(calcium$calcium_1), as.character(calcium$calcium_2)))
write.csv(calcium, file = 'gene_lists/axon_ontologies/condensed/calcium.csv')

dopamin <- as_tibble(read.csv(file = 'gene_lists/axon_ontologies/dopamin.csv', header = T))
dopamin <- unique(c(as.character(dopamin$dopamin_1), as.character(dopamin$dopamin_2)))
write.csv(dopamin, file = 'gene_lists/axon_ontologies/condensed/dopamin.csv')

lysosom <- as_tibble(read.csv(file = 'gene_lists/axon_ontologies/lysosom.csv', header = T))
lysosom <- unique(c(as.character(lysosom$lysosom_1), as.character(lysosom$lysosom_2), as.character(lysosom$lysosom_3), as.character(lysosom$lysosom_4), as.character(lysosom$lysosom_5)))
write.csv(lysosom, file = 'gene_lists/axon_ontologies/condensed/lysosom.csv')

synap <- as_tibble(read.csv(file = 'gene_lists/axon_ontologies/synap.csv', header = T))
synap <- unique(c(as.character(synap$synap_1), as.character(synap$synap_2), as.character(synap$synap_3), as.character(synap$synap_4), as.character(synap$synap_5), as.character(synap$synap_6), as.character(synap$synap_7), as.character(synap$synap_8), as.character(synap$synap_9), as.character(synap$synap_10), as.character(synap$synap_11), as.character(synap$synap_12), as.character(synap$synap_13), as.character(synap$synap_14), as.character(synap$synap_15), as.character(synap$synap_16), as.character(synap$synap_17), as.character(synap$synap_18), as.character(synap$synap_19)))
write.csv(synap, file = 'gene_lists/axon_ontologies/condensed/synap.csv')

endosom <- as_tibble(read.csv(file = 'gene_lists/axon_ontologies/endosom.csv', header = T))
endosom <- unique(c(as.character(endosom$endosom_1), as.character(endosom$endosom_2), as.character(endosom$endosom_3), as.character(endosom$endosom_4), as.character(endosom$endosom_5), as.character(endosom$endosom_6), as.character(endosom$endosom_7), as.character(endosom$endosom_8)))
write.csv(endosom, file = 'gene_lists/axon_ontologies/condensed/endosom.csv')

autophag <- as_tibble(read.csv(file = 'gene_lists/axon_ontologies/autophag.csv', header = T))
autophag <- unique(c(as.character(autophag$autophag_1), as.character(autophag$autophag_2), as.character(autophag$autophag_3), as.character(autophag$autophag_4), as.character(autophag$autophag_5), as.character(autophag$autophag_6), as.character(autophag$autophag_7), as.character(autophag$autophag_8), as.character(autophag$autophag_9), as.character(autophag$autophag_10), as.character(autophag$autophag_11), as.character(autophag$autophag_12), as.character(autophag$autophag_13), as.character(autophag$autophag_14)))
write.csv(autophag, file = 'gene_lists/axon_ontologies/condensed/autophag.csv')

endocyt <- as_tibble(read.csv(file = 'gene_lists/axon_ontologies/endocyt.csv', header = T))
endocyt <- unique(c(as.character(endocyt$endocyt_1), as.character(endocyt$endocyt_2), as.character(endocyt$endocyt_3), as.character(endocyt$endocyt_4), as.character(endocyt$endocyt_5), as.character(endocyt$endocyt_6), as.character(endocyt$endocyt_7), as.character(endocyt$endocyt_8), as.character(endocyt$endocyt_9), as.character(endocyt$endocyt_10), as.character(endocyt$endocyt_11), as.character(endocyt$endocyt_12), as.character(endocyt$endocyt_13), as.character(endocyt$endocyt_14)))
write.csv(endocyt, file = 'gene_lists/axon_ontologies/condensed/endocyt.csv')

clathrin <- as_tibble(read.csv(file = 'gene_lists/axon_ontologies/clathrin.csv', header = T))
clathrin <- unique(c(as.character(clathrin$clathrin_1), as.character(clathrin$clathrin_2), as.character(clathrin$clathrin_3), as.character(clathrin$clathrin_4), as.character(clathrin$clathrin_5), as.character(clathrin$clathrin_6), as.character(clathrin$clathrin_7), as.character(clathrin$clathrin_8), as.character(clathrin$clathrin_9), as.character(clathrin$clathrin_10), as.character(clathrin$clathrin_11), as.character(clathrin$clathrin_12), as.character(clathrin$clathrin_13), as.character(clathrin$clathrin_14)))
write.csv(clathrin, file = 'gene_lists/axon_ontologies/condensed/clathrin.csv')

mito <- as_tibble(read.csv(file = 'gene_lists/axon_ontologies/mito.csv', header = T))
mito <- unique(c(as.character(mito$mito_1), as.character(mito$mito_2), as.character(mito$mito_3), as.character(mito$mito_4), as.character(mito$mito_5), as.character(mito$mito_6), as.character(mito$mito_7), as.character(mito$mito_8), as.character(mito$mito_9), as.character(mito$mito_10), as.character(mito$mito_11), as.character(mito$mito_12), as.character(mito$mito_13), as.character(mito$mito_14), as.character(mito$mito_15), as.character(mito$mito_16), as.character(mito$mito_17), as.character(mito$mito_18), as.character(mito$mito_19), as.character(mito$mito_20), as.character(mito$mito_21), as.character(mito$mito_22), as.character(mito$mito_23), as.character(mito$mito_24), as.character(mito$mito_25), as.character(mito$mito_26), as.character(mito$mito_27), as.character(mito$mito_28), as.character(mito$mito_29), as.character(mito$mito_30), as.character(mito$mito_31), as.character(mito$mito_32), as.character(mito$mito_33), as.character(mito$mito_34), as.character(mito$mito_35), as.character(mito$mito_36), as.character(mito$mito_37), as.character(mito$mito_38), as.character(mito$mito_39), as.character(mito$mito_40), as.character(mito$mito_41), as.character(mito$mito_42), as.character(mito$mito_43), as.character(mito$mito_44), as.character(mito$mito_45), as.character(mito$mito_46), as.character(mito$mito_47)))
write.csv(mito, file = 'gene_lists/axon_ontologies/condensed/mito.csv')

proteas <- as_tibble(read.csv(file = 'gene_lists/axon_ontologies/proteas.csv', header = T))
proteas <- unique(c(as.character(proteas$proteas_1), as.character(proteas$proteas_2), as.character(proteas$proteas_3), as.character(proteas$proteas_4), as.character(proteas$proteas_5), as.character(proteas$proteas_6), as.character(proteas$proteas_7), as.character(proteas$proteas_8), as.character(proteas$proteas_9), as.character(proteas$proteas_10), as.character(proteas$proteas_11), as.character(proteas$proteas_12), as.character(proteas$proteas_13), as.character(proteas$proteas_14), as.character(proteas$proteas_15), as.character(proteas$proteas_16), as.character(proteas$proteas_17)))
write.csv(proteas, file = 'gene_lists/axon_ontologies/condensed/proteas.csv')


```


When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

