---
title: "R Notebook"
output:
  html_document:
    df_print: paged
  pdf_document: default
  html_notebook: default
---

*Test bold text*

Load packages
```{r, results='hide'}
library(DESeq2)
library(tximport)
library(GenomicFeatures)
library(biomaRt)
library(reshape)
library(reshape2)
library(ggplot2)
library(pheatmap)
library(scater)
library(RColorBrewer)
library(dplyr)
library(tidyr)
library(tibble)
library(org.Mm.eg.db)
library(RUVSeq)
library(IHW)
library(Biobase)
library(preprocessCore)
library(EDASeq)
library(fgsea)
```

```{r setup}
dir.create(file.path('.', 'previous'), showWarnings = F)
output_dir <- file.path('output',paste('deseq', Sys.Date(), format(Sys.time(), "%H:%M"), sep = '_'))
dir.create(output_dir, recursive = T)
base_dir <- getwd()
knitr::opts_knit$set(root.dir = output_dir)
```

*Create functions for repeated code*
```{r}
run_deseq <- function() {
  #subset sample metadata
  sample_metadata <- sample_metadata_all %>% data.frame() %>%
                    as_tibble() %>% filter(eval(parse(text=metadata_filter)))
  #create list of kallisto input files
  files <- file.path(kallisto_path, sample_metadata$sample_code, "abundance.h5")
  names(files) <- sample_metadata$sample_name
  
  #import kallisto files
  txi <- tximport(files, type = "kallisto", tx2gene = tx2gene_ERCC, ignoreTxVersion = T)
  
  #create deseq2 object
  dds <- DESeqDataSetFromTximport(txi, colData = sample_metadata, design = design)
  
  #filter low counts
  keep_feature <- rowMeans(counts(dds)) >= min_counts
  dds <- dds[keep_feature, ]
  paste('Dimensions of sample data: ', dim(dds))
  
  #calculate XYZ
  dds <- DESeq(dds, minReplicatesForReplace = replicate_replace)
  assign(paste('dds', name, sep = '_'), envir = .GlobalEnv, dds)
  
  #create contrast variable
  contrast_name <- tail(resultsNames(dds), n=1)
  contrast_split <- strsplit(contrast_name, '_')
  contrast <- c(contrast_split[[1]][1], contrast_split[[1]][2], contrast_split[[1]][4])
  
  #calculate result
  res <- results(dds, alpha = alpha, lfcThreshold = res_lfc, contrast = contrast)
  assign(paste('res', name, sep = '_'), envir = .GlobalEnv, res)
  summary(res)
  
  #perform apeglm shrinkage
  res_shrunken <- lfcShrink(dds, coef = tail(resultsNames(dds), n=1), type = "apeglm", 
                            lfcThreshold = res_lfc, res = res, quiet = T)
  assign(paste('res_shrunken', name, sep = '_'), envir = .GlobalEnv, res_shrunken)
  
  #add independent hypothesis weighting using basemean
  deRes <- as.data.frame(res)
  ihwRes <- ihw(pvalue ~ baseMean,  data = deRes, alpha = alpha)
  resIHW <- results(dds, filterFun=ihw, alpha = alpha, lfcThreshold = res_lfc, 
                    contrast = contrast)
  assign(paste('resIHW', name, sep = '_'), envir = .GlobalEnv, resIHW)
  
  res = resIHW
  
  #put results QC in (pvalue rejections, basemeans)

  #create results output object
  res_tb <- res %>%
    data.frame() %>%
    rownames_to_column(var="gene") %>% 
    as_tibble()

  sig <- res_tb %>%
    filter(padj < alpha & abs(log2FoldChange) > lfc)
  sig <- merge(as.data.frame(sig), anno, by.x = c("gene"), by.y = c("ensembl_gene_id"))
  sig <- merge(as.data.frame(sig), as.data.frame(counts(dds, normalized=TRUE)), 
               by.x = c("gene"), by.y = "row.names", sort=F)
  assign(paste('sig', name, sep = '_'), envir = .GlobalEnv, sig)
  
  res_tb <- merge(as.data.frame(res_tb), anno, by.x = c("gene"), by.y = c("ensembl_gene_id"))
  res_tb <- merge(as.data.frame(res_tb), as.data.frame(counts(dds, normalized=TRUE)), 
                  by.x = c("gene"), by.y = "row.names", sort=F)
  assign(paste('res_output', name, sep = '_'), envir = .GlobalEnv, res_tb)
  
  write.csv(res_tb, file = paste('results', '_', name, '.csv', sep = ''))
}
```

Input folder locations
```{r}
sample_metadata_all <- read.csv(file.path(base_dir, 'sample_metadata_250619.csv'), header = T)
kallisto_path <- "/zfs/analysis/active_r_projects/trap_kall_protcod_ercc/"
```

Load annotation database and mouse ERCC92 tx2gene
```{r}
# txdb <- makeTxDbFromGFF('/zfs/referenceGenomes/mouse/Mus_musculus.GRCm38.96.gtf')
# k <- keys(txdb, keytype = "TXNAME")
# tx2gene <- AnnotationDbi::select(txdb, k, "GENEID", "TXNAME")
# ERCC92 <- read.csv("ERCC92_names.txt", header = F)
# ERCC92$V2 <- ERCC92$V1
# colnames(ERCC92) <- c("TXNAME", "GENEID")
# tx2gene_ERCC <- bind_rows(tx2gene, ERCC92)

load(file.path(base_dir, 'anno_mouse.RData'))
load(file.path(base_dir, 'tx2gene_mouse_ercc.RData'))
```

*Global parameters*
```{r}
alpha = 0.01
res_lfc = 0 #lfc threshold for DESeq2 results functions
lfc = 0 #lfc threshold for post-results filtering genes
replicate_replace = Inf
```

**Comparison 1 -> Axon IP vs Midbrain IP**
```{r}
name = 'axon_ip_vs_mb_ip'

metadata_filter = "ip == 'ip' & batch_trap %in% c('b3')"

design = ~ collection_day + age + axon

min_counts = 100

run_deseq()
```

**Comparison 2 -> Axon UB vs Midbrain UB**
```{r}
name = 'axon_input_vs_mb_input'

metadata_filter = "ip == 'input' & batch_trap %in% c('b2','b3')"

design <- ~ batch_trap + age + axon

min_counts = 100

run_deseq()
```

**Comparison 3 -> Axon IP vs Axon UB**
```{r}
name = 'axon_ip_vs_axon_ub'

metadata_filter = "axon == 'axon' & batch_trap %in% c('b3')"

design <- ~ age + ip

min_counts = 10

run_deseq()
```

**Comparison 4 -> Midbrain IP vs Midbrain UB**
```{r}
name = 'mb_ip_vs_mb_ub'

metadata_filter = "axon == 'mb' & batch_trap %in% c('b2','b3')"

design <- ~ batch_trap + age + ip

min_counts = 10

run_deseq()
```

**Comparison 5 -> Dorsal Striatum IP vs Ventral Striatum IP**
```{r}
name = 'ds_ip_vs_vs_ip'

metadata_filter = "axon == 'axon' & ip == 'ip'"

design <- ~ collection_day + age + region

min_counts = 100

run_deseq()
```

**Comparison 6 -> Dorsal Striatum UB vs Ventral Striatum UB**
```{r}
name = 'ds_ub_vs_vs_ub'

metadata_filter = "axon == 'axon' & ip == 'input'"

design = ~ collection_day + age + region

min_counts = 100

run_deseq()
```

**Comparison 7 -> Dorsal Striatum IP vs Dorsal Striatum UB**
```{r}
name = 'ds_ip_vs_ds_ub'

metadata_filter = "region == 'ds'"

design = ~ age + ip

min_counts = 10

run_deseq()
```

**Comparison 8 -> Ventral Striatum IP vs Ventral Striatum UB**
```{r}
name = 'vs_ip_vs_vs_ub'

metadata_filter = "region == 'vs'"

design = ~ age + ip

min_counts = 10

run_deseq()
```
